src/utils/time.ts

```diff
@@ -1,17 +1,28 @@
-export interface TimeComponents {
-    hours?:number;
-    minutes?:number;
-    seconds?:number;
-    ms?:number;
+enum ConcreteTimeComponents {
+    hours = 0,
+    minutes,
+    seconds,
+    ms
 }
 
-export interface DateComponents {
-    years?:number,
-    months?:number,
-    days?:number,
+export type TimeComponents = {
+    [Property in keyof typeof ConcreteTimeComponents]?: number
 }
 
-export interface DateTimeComponents extends DateComponents, TimeComponents {
+enum ConcreteDateComponents {
+    years = 4,
```

Comment:
Brittle!
`years = Math.max(...Object.values(ConcreteTimeComponents).filter(isFinite) as number[]) + 1,`?

(It is a bit ugly, because enums generate with reverse mappings, such that `Object.keys` and `Object.values` both contain both keys and values. Maybe a comment to that effect:
`// .filter(isFinite) necessary due to reverse mapping elements`,
lest we later forget and erroneously remove the filter? I certainly had forgotten and wondered, why at first I'd always get `[NaN]` as a result.)

Reply:
Oof. Too complicated for my taste. Then I'd rather just explicitly number all of them. Or at this point, just make DateTimeComponents a single interface again..

Reply:
Alas, but fair enough.

I suggest adding a comment after `ms` to this effect ("update `ConcreteDateComponents.years` if adding here") and then just `years = ConcreteTimeComponents.ms + 1,`.

Reply:
Added comments.